mm2src/coins/z_coin.rs

```diff
@@ -793,21 +822,39 @@ impl<'a> UtxoCoinBuilder for ZCoinBuilder<'a> {
         .expect("DEX_FEE_Z_ADDR is a valid z-address");
 
         let z_tx_prover = match &self.z_coin_params.zcash_params_path {
-            None => async_blocking(LocalTxProver::with_default_location)
-                .await
-                .or_mm_err(|| ZCoinBuildError::ZCashParamsNotFound)?,
+            None => {
+                async_blocking(move || {
+                    let params_dir = default_params_folder().or_mm_err(|| ZCoinBuildError::ZCashParamsNotFound)?;
+                    let (spend_path, output_path) = get_spend_output_paths(params_dir)?;
+                    let verification_successful = verify_checksum_zcash_params(
+                        spend_path.clone(),
+                        H256::from(SAPLING_SPEND_EXPECTED_HASH),
+                        output_path.clone(),
+                        H256::from(SAPLING_OUTPUT_EXPECTED_HASH),
+                    )?;
+                    if verification_successful {
+                        Ok(LocalTxProver::new(&spend_path, &output_path))
+                    } else {
+                        MmError::err(ZCoinBuildError::ChecksumVerificationFailed)
+                    }
+                })
+                .await?
+            },
             Some(file_path) => {
                 let path = PathBuf::from(file_path);
                 async_blocking(move || {
-                    let (spend_path, output_path) = if path.exists() {
-                        (path.join(SAPLING_SPEND_NAME), path.join(SAPLING_OUTPUT_NAME))
+                    let (spend_path, output_path) = get_spend_output_paths(path)?;
+                    let verification_successful = verify_checksum_zcash_params(
+                        spend_path.clone(),
+                        H256::from(SAPLING_SPEND_EXPECTED_HASH),
+                        output_path.clone(),
+                        H256::from(SAPLING_OUTPUT_EXPECTED_HASH),
+                    )?;
+                    if verification_successful {
+                        Ok(LocalTxProver::new(&spend_path, &output_path))
                     } else {
-                        return MmError::err(ZCoinBuildError::ZCashParamsNotFound);
-                    };
-                    if !(spend_path.exists() && output_path.exists()) {
-                        return MmError::err(ZCoinBuildError::ZCashParamsNotFound);
+                        MmError::err(ZCoinBuildError::ChecksumVerificationFailed)
```

Comment:
The code in match arms is pretty much duplicated, you can use the following instead:
```rust
        let params_dir = match &self.z_coin_params.zcash_params_path {
            None => default_params_folder().or_mm_err(|| ZCoinBuildError::ZCashParamsNotFound)?,
            Some(file_path) => PathBuf::from(file_path),
        };

        let z_tx_prover = async_blocking(move || {
            let (spend_path, output_path) = get_spend_output_paths(params_dir)?;
            let verification_successful = verify_checksum_zcash_params(
                &spend_path,
                H256::from(SAPLING_SPEND_EXPECTED_HASH),
                &output_path,
                H256::from(SAPLING_OUTPUT_EXPECTED_HASH),
            )?;
            if verification_successful {
                Ok(LocalTxProver::new(&spend_path, &output_path))
            } else {
                MmError::err(ZCoinBuildError::ChecksumVerificationFailed)
            }
        })
        .await?;
```

Reply:
Yeah, it looks better, thanks
done